repository_service_tuf_worker/repository.py

```diff
@@ -398,61 +417,128 @@ def bootstrap(
 
         return asdict(result)
 
-    def add_targets(self, payload: Dict[str, Any], update_state: str) -> None:
+    def publish_targets(self):
         """
-        Updates 'bins' roles metadata, assigning each passed target to the
-        correct bin.
+        Publish targets from SQL DB as persistent TUF Metadata in the backend
+        storage.
+        """
+
+        # lock to avoid race conditions
+        with self._redis.lock("publish_targets", timeout=5.0):
+            # get all delegated role names with unpublished targets
+            unpublished_roles = targets_crud.read_unpublished_rolenames(
+                self._db
+            )
 
-        Assignment is based on the hash prefix of the target file path. All
-        metadata is signed and persisted using the configured key and storage
-        services.
+            if len(unpublished_roles) == 0:
+                logging.debug("No new targets in delegated roles. Finishing")
+                return None
```

Comment:
Maybe you can also release the redis lock before returning so that the lock is not taken for whole 5 seconds.

Reply:
We use context (see line 427)
When it returns the lock is released.

Reply:
Sorry, but I read line 427 `with self._redis.lock("publish_targets", timeout=5.0):` and I don't see where we actually say we use context.
Am I right to guess that at line `427` we lock the `publish_targets` which will be released by another function somewhere else?

Reply:
We use the context `with self._redis.lock()`
Same as we use `with open()` for files. You never close the file. You will be using the context. When the block finishes, automatically close the file. The same applies to the lock. When the block ends, it releases the lock.

Reply:
Got it, thanks.